src/include/txn.i

```diff
@@ -308,9 +308,28 @@ __wt_txn_visible_all(
 #ifdef HAVE_TIMESTAMPS
 	{
 	WT_TXN_GLOBAL *txn_global = &S2C(session)->txn_global;
-	int cmp;
+	int cmp, stable_cmp;
 
 	/* Timestamp check. */
+        /*
+         * If we are given a timestamp but no oldest or stable timestamp
+         * has been set then the timestamp is not visible.  Otherwise
+         * anything older than the oldest timestamp and newer than the
+         * stable timestamp is not visible.
+         */
+        if (timestamp != NULL) {
+                if (!txn_global->has_stable_timestamp ||
+                    !txn_global->has_oldest_timestamp)
+                        return (false);
+	        __wt_readlock(session, &txn_global->rwlock);
+	        stable_cmp = __wt_timestamp_cmp(timestamp,
+                    &txn_global->stable_timestamp);
+	        cmp = __wt_timestamp_cmp(timestamp,
+                    &txn_global->oldest_timestamp);
+	        __wt_readunlock(session, &txn_global->rwlock);
+                if (stable_cmp > 0 || cmp < 0)
+                        return (false);
+        }
```

Comment:
This seems overly complicated because the `pinned_timestamp` has to be less than either the stable or oldest timestamp and we're already checking for that below.

I *think* all we need here is:

```
if (timestamp == NULL)
        return (true);
else if (!txn_global->has_pinned_timestamp)
        return (false);
```

Does that make sense?